torch/testing/_internal/optests/compile_check.py

```diff
@@ -0,0 +1,98 @@
+import torch
+from .make_fx import make_fx_check
+from .aot_autograd import aot_autograd_check
+from torch._subclasses.schema_check_mode import SchemaCheckMode
+
+
+def operator_compile_check(
+        func,
+        args,
+        kwargs=None,
+        *,
+        dynamic_only=False,
+        supports_autograd=True,
+        fullgraph=True,
+):
+    """Check if torch.compile supports a custom operator.
+
+    Args:
+        func (function): an operator that takes at least one Tensor as input
+            and returns a Tensor or a Tuple of Tensors.
+        args (Tuple): args to the operator
+        kwargs (dict, optional): kwargs to the operator
+        dynamic_only (bool, optional): If the operator only works with dynamic
+            shapes. This can happen if it returns Tensors whose shape are
+            dependent on the data on the input Tensors. If True, we skip
+            tests related to torch.compile with static shapes.
+        supports_autograd (bool, optional): If the operator does not support
+            autograd. If False, we will skip autograd-related tests.
+        fullgraph (bool, optional): If we expect torch.compile to not graph
+            break on seeing the operator. Graph breaks are bad for performance.
+
+    """
+    if kwargs is None:
+        kwargs = {}
+
+    # Add some clones to the beginning of func so that it is actually out-of-place.
+    # A lot of our tests (e.g. aot_autograd_check) assume that the function passed
+    # in is out-of-place and contains a call to the operator in being tested.
```

Comment:
mostly as a confirmation/reminder for me: `CustomOp` doesn't allow you to create ops with mutable schemas, so for `CustomOp`, these clones shouldn't be necessary. Is this API also meant to be used with existing C++ custom ops, which might perform views and mutations?

Reply:
> Is this API also meant to be used with existing C++ custom ops, which might perform views and mutations?

Yes

Reply:
Hmm - totally reasonable if you think this isn't worth doing, but: given that AOTAutograd has to do a whole bunch of extra work when mutation and aliasing comes into the picture, do you think it's worth exercising that logic more directly in the custom op tests?

Specifically: Written as is (where we clone all tensor inputs), we'll never end up invoking AOTAutograd's custom logic around "what happens when you do an input mutation, or return an output that aliases an input". `test/functorch/test_aotdispatch.py` has a bunch of tests that try to explicitly handle this case, by:

(1) running functions that cause input mutations, and checking that the inputs match compared to eager

(2) run functions that return aliases of inputs, explicitly mutate the output afterwards, and check that inputs match (and also confirm that the computed gradients match

One argument for *not* doing this that I think is fair: If that extra logic to fail, that would be more indicative of an AOTAutograd bug, than a bug in your custom op implementation. Should `compile_check()` be playing that role?

Then again, we're also running inductor directly as part of this check, so maybe we do just want to exercise as much of the system as possible.

Reply:
"custom C++ op that returns an alias of an input, but has a custom autograd rule" is probably a good example of something that this could catch (aka https://github.com/pytorch/pytorch/issues/102970)

Reply:
Is this something you think we would/should add to all of the OpInfo tests for AOTAutograd?

Reply:
hmm. I think it probably isn't worth adding to all the OpInfo tests, since this doesn't matter for 99% of ops (and that would probably increase test time quite a bit). I guess I'm just a bit more paranoid about custom ops, since we've seen custom ops with wilder schemas than most of our ATen ops.

Reply:
IMO it's good to add more tests. Would it be fine if I took this as a follow-up item? compile_check as written is useful today, not public yet, and figuring out where to check input mutations is a bit complicated

Reply:
totally fine!